### PR TITLE
stats: aggregated metrics

### PR DESCRIPTION
**Pull request by alissa-tung**

### CODE REVIEW

File: api/v1/stats/service.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+	if len(allStatsRaw) != 0 {
+		headers := allStatsRaw[0].Headers
+
+		allStats := model.TableResult{
+			Headers: headers,
+			Value:   make([]map[string]string, 0),
+		}
+
+		for _, stats := range allStatsRaw {
```
Comment (YangKian) [inline api/v1/stats/service.go:136]: Can you explain the logic of aggregation with some real examples? It seems that this implementation may contain some bugs.

File: api/v1/stats/service.go [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+	} else {
+		var err errorno.ErrorResponse
+		if len(allStatsRaw) != 0 {
+			err = adminError(fmt.Errorf("got stats of length 0"))
+		} else if len(allStatsRaw[0].Value) != 1 {
+			err = adminError(fmt.Errorf("the length value should be 1"))
+		}
+		c.AbortWithStatusJSON(http.StatusInternalServerError, err)
+		return
+	}
```
Comment (YangKian) [inline api/v1/stats/service.go:209]: ditto, Use some short judgments to quickly short-circuit the wrong scenarios
``` go
if len(allStatsRaw) != 0 { return ...}
if len(allStatsRaw[0].Value) != 1 { return ...}

// handle normal scenario
```

**Review by YangKian — CHANGES_REQUESTED** [review]

File: api/v1/stats/service.go (hunk 2 above)
Comment (alissa-tung) [reply api/v1/stats/service.go]: thx, [22ba412](https://github.com/hstreamdb/http-services/pull/12/commits/22ba4124c1076eaa0e7afcca97cc129ca80902ad)

File: api/v1/stats/service.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+				}
+				stats.Value = append(stats.Value, mp)
+			}
+
+			allStatsRaw = append(allStatsRaw, stats)
+			return nil
+		}()
+		if err != nil {
+			return nil, err
+		}
```
[the remark this answers is not in the document]
Comment (alissa-tung) [reply api/v1/stats/service.go:115]: sendAdminRequest does send request to random server, in the go-client pkg side, it calls serverinfo and select one url. to reuse the logic, maybe i should file a pr to go-client pkg first?

File: internal/rpcService/admin.go [hunk 4]
```diff
@@ -22,21 +22,35 @@ func (c *HStreamClient) GetStats(category string, metrics string, intervals []st
...
+func (c *HStreamClient) sendAdminRequestToServer(addr, cmd string) (*model.TableType, error) {
+	resp, err := c.client.AdminRequestToServer(addr, cmd)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.parseAdminRequest(resp)
+}
+
+func (c *HStreamClient) parseAdminRequest(resp string) (*model.TableType, error) {
```
Comment (YangKian) [inline internal/rpcService/admin.go:45]: rename this function to `parseAdminResponse`

File: api/v1/stats/service.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+				mainKey := val[headers[0]]
+
+				ok := false
+				ix := -1
+				for i, cur := range allStats.Value {
+					if cur[mainKey] == val[mainKey] {
+						ok = true
+						ix = i
+					}
+				}
```
Comment (YangKian) [inline api/v1/stats/service.go:129]: If you just want to find the idx in allStats.Value which `allStats.Value[idx][mainKey] == val[mainKey]`, why you don't break the loop when you find the idx?

Comment (alissa-tung) [reply api/v1/stats/service.go]: thx! fixed in [7bc88a4](https://github.com/hstreamdb/http-services/pull/12/commits/7bc88a4c758dc7bf50dbfeff1d1170155097aa93)

**Pull request merged**